ci-operator/step-registry/cucushift/upgrade/rollback/cucushift-upgrade-rollback-commands.sh

```diff
@@ -0,0 +1,159 @@
+    do
+        echo >&2 "Failed to extract oc binary, retry..."
+        (( retry -= 1 ))
+        if (( retry < 0 )); then return 1; fi
+        sleep 60
+    done
+    mv ${tmp_oc}/oc ${OC_DIR} -f
+    export PATH="$PATH"
+    which oc
+    oc version --client
+    return 0
+}
+
+function run_command() {
+    local CMD="$1"
+    echo "Running command: ${CMD}"
+    eval "${CMD}"
+}
+
+# Rollback the cluster to target release
+function rollback() {
+    run_command "oc adm upgrade rollback"
+    echo "Rolling back cluster gets started..."
+}
+
+# Monitor the rollback status
+function check_rollback_status() {
+    local wait_rollback="${TIMEOUT}" out avail progress cluster_version
+    cluster_version="${TARGET_VERSION}"
+    echo "Starting the rollback checking on $(date "+%F %T")"
+    while (( wait_rollback > 0 )); do
+        sleep 5m
+        wait_rollback=$(( wait_rollback - 5 ))
+        if ! ( run_command "oc get clusterversion" ); then
+            continue
+        fi
+        if ! out="$(oc get clusterversion --no-headers || false)"; then
+            echo "Error occurred when getting clusterversion"
+            continue 
+        fi
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ ${avail} == "True" && ${progress} == "False" && ${out} == *"Cluster version is ${cluster_version}" ]]; then
+            echo -e "Rollback succeed on $(date "+%F %T")\n\n"
+            return 0
+        fi
+    done
+    if [[ ${wait_rollback} -le 0 ]]; then
+        echo -e "Rollback timeout on $(date "+%F %T"), exiting\n" && return 1
+    fi
+}
+
+# Check version, state in history
+function check_history() {
+    local version state
+    version=$(oc get clusterversion/version -o jsonpath='{.status.history[0].version}')
+    state=$(oc get clusterversion/version -o jsonpath='{.status.history[0].state}')
+    if [[ ${version} == "${TARGET_VERSION}" && ${state} == "Completed" ]]; then
+        echo "History check PASSED, cluster is now rollbacked to ${TARGET_VERSION}" && return 0
+    else
+        echo >&2 "History check FAILED, cluster rollbacked to ${TARGET_VERSION} failed, current version is ${version}, exiting" && return 1
+    fi
+}
+
+if [[ -f "${SHARED_DIR}/kubeconfig" ]] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+fi
+
+# Setup proxy if it's present in the shared dir
+if [[ -f "${SHARED_DIR}/proxy-conf.sh" ]]; then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+# Target version oc will be extract in the /tmp/client directory, use it first
+mkdir -p /tmp/client
+export OC_DIR="/tmp/client"
+export PATH=${OC_DIR}:$PATH
+
+run_command "oc get machineconfig"
+
+export TARGET="${OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE}"
+TARGET_VERSION="$(env "NO_PROXY=*" "no_proxy=*" oc adm release info "${TARGET}" --output=json | jq -r '.metadata.version')"
+extract_oc
```

Comment:
so in this case, extract_oc will extract ${TARGET} version, which is older than the current cluster. wonder if this is intended behavior

Reply:
Yes, I have the same concern. Let me confirm with Trevor.

Reply:
Trevor guesses newer oc would be better. So I've switched to use target-versioned oc

Reply:
ah, good catch!